client/components/swipeable/index.js

```diff
@@ -0,0 +1,249 @@
+import classnames from 'classnames';
+import { useRtl } from 'i18n-calypso';
+import { Children, useState, useEffect, useRef } from 'react';
+
+import './style.scss';
+
+export const Swipeable = ( {
+	hasDynamicHeight = false,
+	children,
+	currentPage = 0,
+	onPageSelect,
+	pageClassName,
+	containerClassName,
+} ) => {
+	const [ pageWidth, setPageWidth ] = useState( null );
+	const [ swipeableArea, setSwipeableArea ] = useState();
+	const isRtl = useRtl();
+
+	const [ pagesStyle, setPagesStyle ] = useState( {
+		transform: `translate3d(0px, 0px, 0px)`,
+		transitionDuration: `300ms`,
+	} );
+	const [ dragStartData, setDragStartData ] = useState();
+
+	const pagesRef = useRef();
+	const resizeObserverRef = useRef();
+	const numPages = Children.count( children );
+
+	const getWidth = () => {
+		return resizeObserverRef.current?.getBoundingClientRect().width;
+	};
+
+	const getPagesWidth = () => {
+		return pageWidth * numPages;
+	};
+
+	const getOffset = ( index ) => {
+		const offset = pageWidth * index;
+		return isRtl ? offset : -offset;
+	};
+
+	function useUpdateLayout( enabled, currentPageIndex, updateLayout ) {
+		// save callback to a ref so that it doesn't need to be a dependency of other hooks
+		const savedUpdateLayout = useRef();
+		useEffect( () => {
+			savedUpdateLayout.current = updateLayout;
+		}, [ updateLayout ] );
+
+		// fire when the `currentPageIndex` parameter changes
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
+			}
+
+			savedUpdateLayout.current();
+		}, [ enabled, currentPageIndex ] );
+
+		// fire when the window resizes
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
+			}
+
+			const onResize = () => savedUpdateLayout.current( false );
+			window.addEventListener( 'resize', onResize );
+			window.addEventListener( 'orientationchange', onResize );
+			return () => {
+				window.removeEventListener( 'resize', onResize );
+				window.removeEventListener( 'orientationchange', onResize );
+			};
+		}, [ enabled ] );
+	}
+
+	const updateEnabled = hasDynamicHeight && numPages > 1;
+
+	useUpdateLayout( updateEnabled, currentPage, ( hasTransitionDuration = true ) => {
+		const offset = getOffset( currentPage );
+		const transform = {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: hasTransitionDuration ? `300ms` : `0ms`,
+		};
+
+		let height = { height: null };
+		const targetHeight = pagesRef.current?.querySelector( '.is-current' )?.offsetHeight;
+
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+		setPageWidth( getWidth() );
+		setSwipeableArea( pagesRef.current?.getBoundingClientRect() );
```

Comment:
I understand that these variables (that are expensive to compute) and now only recalculated when this callback is run by `useUpdatedLayout` — but does it mean that, is `updateEnabled` is `false`, these values never get computed?

Reply:
@sgomes I wonder if you also have any suggestion for those `getBoundingClientRect` calls — I know they can be quite expensive to recalculate.

I initially suggested ([here](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r760468419) and [here](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r760457768)) to memoize them, but in the meantime they have been moved to this callback.

Do you think this change is reasonable?

Reply:
Happy to help! This needs a little bit of extra background to explain thoroughly.

The browser rendering pipeline is roughly divided into three stages: Layout, Paint, and Composite. Very simply put, the first stage determines the general layout of the page in terms of which elements go where, the second stage renders everything into layers, and the final stage takes those layers and combines them into a single layer to be rendered on the user's screen.

React sits atop of this, having its own "Render" stage, which essentially determines the changes to be applied to the real browser DOM before any of the above begins.

The challenge is that `getBoundingClientRect` triggers a layout pass, which, if it happens at the wrong moment, can throw the whole pipeline out of whack, essentially restarting it and delaying rendering beyond the difficult to hit 13ms (1000ms/60fps) target. Namely, we need to avoid taking these measurements as we're getting ready to render on screen.

It's not currently clear to me at what stage this hook is getting called, nor what's driving it, but the general approach inside of it seems perfectly reasonable to me, and follows what is generally the best practice: first, take all your measurements with the potentially heavy layout-hitting methods you need, and then start the animation. If this is done well, then even if you miss some frames, then no animation will have started yet, so it isn't visible to the user or in any way problematic. That is, you make sure to only run Layout as part of the initial setup before you start animating, and then just let the animation run on its own; the animation in question here is a `transform`, which is a CSS property that only triggers the Composite phase on every frame, which is exactly what we want!

As such, assuming that this is being called once, at the beginning of the process, and that it's not being called at the wrong time in the rendering pipeline, this approach should work well. I'd need to spend a bit of time refreshing my React hook knowledge in order to figure out whether this is the case, though 😕

Reply:
Thank you for the rendering pipeline refreshed, Sergio!

I'll let @enejb go through the other comments in this PR (which may also affect how this callback is run), and we may resume this conversation later

Reply:
Thanks for the explenation @sgomes. 
 I tested this with FF and Chrome using 6x CPU slowing and I didn't notice I am still seeing about 58fps any difference between when we did have a useCallback and not. As far as I can tell they don't make any difference. 

I am leaving the change in here. So it might be easier to compare :) if someone else wants to try.

Reply:
Thank you @sgomes and @enejb , we can probably then mark this conversation resolved.